x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/PinnedQueryBuilder.java

```diff
@@ -0,0 +1,218 @@
+import org.apache.lucene.search.MatchNoDocsQuery;
+import org.apache.lucene.search.Query;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.ParsingException;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.lucene.search.Queries;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.index.mapper.IdFieldMapper;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.query.AbstractQueryBuilder;
+import org.elasticsearch.index.query.MatchNoneQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryRewriteContext;
+import org.elasticsearch.index.query.QueryShardContext;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+import static org.elasticsearch.common.xcontent.ObjectParser.fromList;
+
+/**
+ * A query that will promote selected documents (identified by ID) above matches produced by an "organic" query. In practice, some upstream
+ * system will identify the promotions associated with a user's query string and use this object to ensure these are "pinned" to the top of
+ * the other search results.
+ */
+public class PinnedQueryBuilder extends AbstractQueryBuilder<PinnedQueryBuilder> {
+    public static final String NAME = "pinned";
+
+    private static final ParseField IDS_FIELD = new ParseField("ids");
+    public static final ParseField ORGANIC_QUERY_FIELD = new ParseField("organic");
+
+    private final Set<String> ids = new LinkedHashSet<>();
+    private QueryBuilder organicQuery;
+
+    // Organic queries will have their scores capped to this number range,
+    private static final float MAX_ORGANIC_SCORE = Float.intBitsToFloat((0xfe << 23)) - 1;
+
+    /**
+     * Creates a new PinnedQueryBuilder
+     */
+    public PinnedQueryBuilder() {
+        // nothing to do
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public PinnedQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        Collections.addAll(ids, in.readStringArray());
+        organicQuery = in.readNamedWriteable(QueryBuilder.class);
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeStringArray(ids.toArray(new String[ids.size()]));
+        out.writeNamedWriteable(organicQuery);
+    }
+
+    /**
+     * Adds pinned ids to the query.
+     */
+    public PinnedQueryBuilder addIds(String... ids) {
+        if (ids == null) {
+            throw new IllegalArgumentException("[" + NAME + "] ids cannot be null");
+        }
+        Collections.addAll(this.ids, ids);
```

Comment:
should we complain if we are seeing the same id twice?

Reply:
separately I wonder whether we should remove setters and set both ids and the organic query in the constructor?

Reply:
Also should we enforce a limit on the maximum number of pinned hits? It should be less than 1024 to not hit the BooleanQuery max clause count, but we should probably set it ever lower, maybe something like 100?

Reply:
>Also should we enforce a limit on the maximum number of pinned hits?

I wonder if users may over-supply the set of IDs if they know that faceting choices in the UI will act as a filter on the IDs listed in the query? It may be less inefficient to do so, but I could see a scenario where the set of IDs associated with a query string is fixed and held in an upstream system and then passed through to elasticsearch along with criteria from any faceting selections e.g shoe size etc. The 1024 limit would obviously still apply.

Reply:
Good point. I still have a slight bias towards lower bounds given that it is easier to increase such limits in the future than decrease them. Furthermore 1024 is going to move from a per-BooleanQuery limit to a limit on the entire set of clauses of the entire query in Lucene 9.0, so I'd rather like a more conservative limit.

Reply:
I went with a hardcoded limit of 100 for now